File: data-pipeline-libraries-java/data-pipeline-core-java/src/main/java/com/enrichmeai/culvert/dataquality/ViolationKind.java
package com.enrichmeai.culvert.dataquality;

/**
 * Classification of a single field violation found during data-quality validation.
 *
 * <ul>
 *   <li>{@link #MISSING_REQUIRED} — a field declared {@code REQUIRED} in the
 *       {@code EntitySchema} had a {@code null} or absent value.</li>
 *   <li>{@link #TYPE_MISMATCH} — the field's runtime value is not assignable to
 *       the Java type that corresponds to the schema wire type
 *       ({@code STRING→String}, {@code INT64→Long}, {@code FLOAT64→Double},
 *       {@code BOOL→Boolean}).</li>
 *   <li>{@link #OUT_OF_RANGE} — the field's numeric value falls outside the
 *       {@link NumericRange} that was supplied to
 *       {@link DataQualityTransform} at construction time.</li>
 * </ul>
 *
 * <p>Zero GCP / Beam dependencies — this enum lives in the cloud-neutral core.
 *
 * @since Sprint 14 / issue #73
 */
public enum ViolationKind {

    /** The field is declared REQUIRED but its value is null or missing. */
    MISSING_REQUIRED,

    /** The field's runtime value type does not match the schema-declared wire type. */
    TYPE_MISMATCH,

    /** The field's numeric value falls outside its declared {@link NumericRange}. */
    OUT_OF_RANGE
}

File: data-pipeline-libraries/data-pipeline-core/src/data_pipeline_core/dataquality/violation_kind.py
"""ViolationKind — classification of a single field violation.

Mirrors ``com.enrichmeai.culvert.dataquality.ViolationKind`` (Java
Sprint 14 / issue #73).  Three variants, same names, same semantics.

Port: T18.2 / issue #118.
"""

from __future__ import annotations

from enum import Enum


class ViolationKind(Enum):
    """Classification of a single field violation found during validation.

    Mirrors ``ViolationKind.java`` (see
    ``data-pipeline-libraries-java/.../dataquality/ViolationKind.java``):

    * ``MISSING_REQUIRED`` — a field declared ``REQUIRED`` had a ``None``
      or absent value. (Java line 25)
    * ``TYPE_MISMATCH``    — the field's runtime type does not match the
      schema-declared wire type. (Java line 28)
    * ``OUT_OF_RANGE``     — a numeric field falls outside its declared
      :class:`NumericRange`. (Java line 31)
    """

    MISSING_REQUIRED = "MISSING_REQUIRED"
    """The field is declared REQUIRED but its value is None or missing."""

    TYPE_MISMATCH = "TYPE_MISMATCH"
    """The field's runtime value type does not match the schema wire type."""

    OUT_OF_RANGE = "OUT_OF_RANGE"
    """The field's numeric value falls outside its declared NumericRange."""
